content/posts/reviewing-unsafe-rust.md

```diff
@@ -0,0 +1,1411 @@
+where
+    E: 'static,
+{
+    if TypeId::of::<E>() == target {
+        // Caller is looking for an E pointer and e is ErrorImpl<E>, take a
+        // pointer to its E field.
+        let unerased = e as *const ErrorImpl<()> as *const ErrorImpl<E>;
+        let addr = &(*unerased)._object as *const E as *mut ();
+        Some(NonNull::new_unchecked(addr))
+    } else {
+        None
+    }
+}
+
+// (line 582)
+// Safety: requires layout of *e to match ErrorImpl<ContextError<C, E>>.
+#[cfg(feature = "std")]
+unsafe fn context_downcast<C, E>(e: &ErrorImpl<()>, target: TypeId) -> Option<NonNull<()>>
+where
+    C: 'static,
+    E: 'static,
+{
+    if TypeId::of::<C>() == target {
+        let unerased = e as *const ErrorImpl<()> as *const ErrorImpl<ContextError<C, E>>;
+        let addr = &(*unerased)._object.context as *const C as *mut ();
+        Some(NonNull::new_unchecked(addr))
+    } else if TypeId::of::<E>() == target {
+        let unerased = e as *const ErrorImpl<()> as *const ErrorImpl<ContextError<C, E>>;
+        let addr = &(*unerased)._object.error as *const E as *mut ();
+        Some(NonNull::new_unchecked(addr))
+    } else {
+        None
+    }
+}
+
+// (line 626)
+// Safety: requires layout of *e to match ErrorImpl<ContextError<C, Error>>.
+unsafe fn context_chain_downcast<C>(e: &ErrorImpl<()>, target: TypeId) -> Option<NonNull<()>>
+where
+    C: 'static,
+{
+    if TypeId::of::<C>() == target {
+        let unerased = e as *const ErrorImpl<()> as *const ErrorImpl<ContextError<C, Error>>;
+        let addr = &(*unerased)._object.context as *const C as *mut ();
+        Some(NonNull::new_unchecked(addr))
+    } else {
+        // Recurse down the context chain per the inner error's vtable.
+        let unerased = e as *const ErrorImpl<()> as *const ErrorImpl<ContextError<C, Error>>;
+        let source = &(*unerased)._object.error;
+        (source.inner.vtable.object_downcast)(&source.inner, target)
+    }
+}
+```
+
+The basic `object_downcast()` function seems fairly reasonable, if the type
+we're trying to downcast to (`target`) matches the type this vtable was created
+for (`E`), the type check passes and we can return a pointer to `_object`.
+
+I also really appreciate this comment:
+
+> ```rust
+> // Caller is looking for an E pointer and e is ErrorImpl<E>, take a
+> // pointer to its E field.
+> ```
+
+While it's nothing we don't already know after having read through most of
+the `error.rs` file in great detail, it's really useful to reiterate your
+assumptions for someone who doesn't have that background knowledge.
+
+Something that jumped out at me was this line:
+
+```rust
+let unerased = e as *const ErrorImpl<()> as *const ErrorImpl<E>;
+let addr = &(*unerased)._object as *const E as *mut ();
+```
+
+We're starting off with an immutable reference (`&ErrorImpl<E>`) then borrowing
+the `_object` field (immutably) and using pointer casts to get rid of the
+`const`.
```

Comment:
The cast from \*const to \*mut is required for using NonNull, whose constructors take \*mut but is intended for use with both \*const and \*mut. There isn't a separate NonNull type dedicated to \*const.

Reply:
I recently discovered that `NonNull<T>` implements `From<&T>` and `From<&mut T>`. So maybe the overall `object_downcast()` function could look like this?

```rust
unsafe fn object_downcast<E>(e: NonNull<ErrorImpl<()>>, target: TypeId) -> Option<NonNull<()>> where E: 'static,
{
    if TypeId::of::<E>() == target {
        let unerased = e.cast::<ErrorImpl<E>>().as_ref();
        Some(NonNull::from(&unerased._object))
    } else {
        None
    }
}
```